### PR TITLE
Add a grouped bnb 4-bit training forward for gpt-oss experts

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

The gpt-oss 4-bit training path loops over all experts per layer, launching a dequantize and two matmuls per expert even for experts that received no tokens. A profiled step is about 80 percent GPU-idle: 2.27s of CPU work against 0.61s of CUDA, with 33k kernel launches, 2062 per-expert 4-bit matmuls, and 3140 dequantize calls.

Stacked on #838 (the base is `moe-recompute-backward`, which provides the grouped matmul primitives); the diff here is only the gpt-oss commit.

## What this does

Adds a grouped equivalent used when every expert is a plain LoRA-free `Linear4bit` with a populated quant_state: routing indices are sorted once on GPU, each projection stack is dequantized with a single `dequantize_4bit` call over one cross-expert QuantState (packed NF4 bytes are elementwise row-major, so concatenating per-expert bytes and fp32 absmax is exact), and the two projections run as `torch._grouped_mm` calls with grouped bias adds, followed by the identical fp32 index_add combine. With `UNSLOTH_MOE_RECOMPUTE=1` the dequantized stacks are rebuilt in backward instead of staying on the tape.

The gate lives in `torch_native_forward` (the function the experts class actually dispatches) behind a readiness probe (grouped_mm support, no LoRA wrappers, frozen quantized weights), can be disabled with `UNSLOTH_GPTOSS_GROUPED=0`, and falls back to the per-expert loop on any failure.

## Testing

- gpt-oss-20b-unsloth-bnb-4bit LoRA SFT at 4k tokens: 10194 tok/s grouped vs 2433 with the loop (4.2x); 11238 tok/s at 8k.
- Per-step losses match the loop path to 0.0022 over 20 deterministic steps.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
The grouped path now requires every expert bias to be present, since a missing bias would break the `torch.stack`; otherwise it falls back to the per-expert loop, which handles a None bias. It also returns float32 during training to match the native fallback's fp16 NaN protection. Verified on a live grouped-mm run.

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 1]
```diff
@@ -848,13 +849,133 @@ def __init__(self, config):
...
+        def _stack(projs, quant_state):
+            # One fused dequant gives (E, out, in); grouped_mm takes the transposed view.
+            data = torch.cat([l.weight.data.reshape(-1, 1) for l in projs])
+            return bnb.functional.dequantize_4bit(data, quant_state).transpose(1, 2)
+
+        xg = hidden_states[sorted_tokens]
+        gate_up = _base_grouped_mm(
+            xg, offsets, lambda: _stack(self.gate_up_projs, gate_up_qs), recompute)
+        gate_up = gate_up + gate_up_bias[expert_ids]
+        gated = swiglu_torch_forward(gate_up, self.alpha, self.limit)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py:993]: swiglu_torch_forward already computes in float32 internally and clamps to the limit, and torch grouped_mm needs matching operand dtypes, so forcing a float32 gated output would make every grouped call raise and permanently fall back to the loop. grouped_mm capable hardware is bf16 capable and gpt_oss requests fp16 through forced float32 mixed precision, so the fp16 scenario does not reach this path.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 2]
```diff
@@ -848,13 +849,138 @@ def __init__(self, config):
...
+        out = _base_grouped_mm(
+            gated, offsets, lambda: _stack(self.down_projs, down_qs), recompute)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py:995]: Fair point on the protection asymmetry. fp16 is not reachable through the loader (gpt_oss is in FORCE_FLOAT32, so fp16 requests load bf16), but the grouped dispatch now also gates on dtype != fp16 at both sites, so an fp16 run keeps the loop path with the fp32 swiglu and autocast-disabled down projection. Kept the grouped mm in a single dtype since mixing fp32 into it would fail dtype matching.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 3]
```diff
@@ -848,13 +849,144 @@ def __init__(self, config):
             "down_proj_bias", torch.empty(0, dtype=self.dtype), persistent=False
         )
 
+    def _grouped_bnb4bit_ready(self):
+        """True when every expert is a plain (LoRA-free) bnb Linear4bit with a
+        populated quant_state, so the grouped torch._grouped_mm path is exact."""
+        if os.environ.get("UNSLOTH_GPTOSS_GROUPED", "1") == "0":
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py:860]: The grouped members are now self-contained (local imports, no module globals), so a source-copied class in the standalone compiled cache cannot hit NameError. Verified by exec-ing the class source in a bare namespace and running the probe.

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 4]
```diff
@@ -848,13 +849,144 @@ def __init__(self, config):
...
+                    return (absmax + qs.offset).float()
+                return qs.absmax.float()
+
+            def build_qs(projs):
+                states = [l.weight.quant_state for l in projs]
+                absmax = torch.cat([_absmax_fp32(qs) for qs in states])
+                q0 = states[0]
+                return QuantState(
+                    absmax=absmax,
+                    shape=torch.Size((len(projs),) + tuple(q0.shape)),
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py:964]: Readiness now requires every expert to tile into whole blocks of one shared blocksize before the packed byte and absmax concat, else it falls back to the per-expert loop. Production gpt-oss shapes are exact multiples, but the guard makes it structural.

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 5]
```diff
@@ -848,13 +849,144 @@ def __init__(self, config):
...
+                torch.stack([l.bias for l in self.gate_up_projs]),  # (E, 2I)
+                torch.stack([l.bias for l in self.down_projs]),     # (E, H)
+            )
+            self._unsloth_grouped_qs = cached
+        gate_up_qs, down_qs, gate_up_bias, down_bias = cached
+
+        def _stack(projs, quant_state):
+            # One fused dequant gives (E, out, in); grouped_mm takes the transposed view.
+            data = torch.cat([l.weight.data.reshape(-1, 1) for l in projs])
+            return bnb.functional.dequantize_4bit(data, quant_state).transpose(1, 2)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py]: The dequantized stacks are now cast to the input dtype, so fp32 hidden states run grouped_mm directly instead of raising and falling back. Verified on GPU with fp32 inputs.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
### Benchmark: adaptive gradient-checkpoint-aware recompute (before vs after)

Quantified the adaptive `_moe_recompute_default` policy on the real grouped 4-bit expert path.

**Setup:** `unsloth/gpt-oss-20b-unsloth-bnb-4bit`, attention-only LoRA, bf16, `per_device_train_batch_size=2` x `gradient_accumulation_steps=3`, `seq=2048`, 61 steps, single B200. The grouped path engaged in every run (17568 `_base_grouped_mm` calls each). "Before" is the two static choices via `UNSLOTH_MOE_RECOMPUTE` (`=1` always recompute, `=0` always pin); "after" is the adaptive default (unset). Each regime's trio was run on one GPU for clean deltas, with an adaptive-vs-itself replicate to establish the run-to-run noise floor (**0.5%**). Metrics are the mean over post-warmup steps; peak GB is `max_memory_allocated` across steps.

**Gradient checkpointing ON (the default)**

| Variant | Peak GB | Step ms | tok/s |
|---------|---------|---------|-------|
| always-recompute (`=1`, before) | 16.40 | 1908 | 6439 |
| **adaptive (unset, after)** | 16.88 | **1829** | **6718** |
| adaptive replicate | 16.88 | 1839 | 6682 |
| always-pin (`=0`) | 16.88 | 1825 | 6734 |

**Gradient checkpointing OFF**

| Variant | Peak GB | Step ms | tok/s |
|---------|---------|---------|-------|
| always-recompute (`=1`) | 16.74 | 2691 | 4566 |
| **adaptive (unset, after)** | 16.74 | 2686 | 4574 |
| adaptive replicate | 16.74 | 2672 | 4598 |
| always-pin (`=0`, before) | 17.23 | 2477 | 4961 |

**Deltas (both well above the 0.5% noise floor):**

- With gradient checkpointing: adaptive is **+4.1% faster** than always-recompute (6718 vs 6439 tok/s) and matches always-pin (within 0.2%), at a cost of **+0.48 GB** (the momentary pinned stack).
- Without gradient checkpointing: adaptive stays on the lean path, **16.74 GB, identical to always-recompute**, and avoids always-pin's **+0.49 GB**.

So the adaptive default is never worse than the better static choice in each regime: it takes pin's speed when the pin is momentary (inside a checkpoint recompute pass) and recompute's memory safety otherwise, with no flag needed.

**Mechanism** (probe of the recompute decision at each grouped call):

- GC on, adaptive: the forward pass returns `recompute=True` (moot, it runs under `no_grad`), and the backward re-run returns `recompute=False` (pins). Split was exactly 50/50.
- GC off, adaptive: every call returns `recompute=True` (recompute).

**Loss is numerically transparent.** The cross-variant final-loss spread is within the identical-code noise floor (adaptive vs its own replicate: `max_loss_diff` 0.0144 with GC, 0.245 without), so the spread is bf16 `_grouped_mm` atomic nondeterminism, not the recompute choice. `UNSLOTH_MOE_RECOMPUTE` remains a manual override (`1` force recompute, `0` force pin).

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 6]
```diff
@@ -848,13 +849,133 @@ def __init__(self, config):
...
+                if hasattr(lin, "lora_A") or hasattr(lin, "base_layer"):
+                    return _fail(f"LoRA-wrapped expert {type(lin).__name__}")
+                w = getattr(lin, "weight", None)
+                if not (isinstance(w, Params4bit) and getattr(w, "quant_state", None) is not None):
+                    return _fail(f"expert weight {type(w).__name__} without quant_state")
+                if w.requires_grad:
+                    return _fail("expert weight requires_grad")
+                b = getattr(lin, "bias", None)
+                if b is not None and b.requires_grad:
+                    return _fail("expert bias requires_grad")
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py:913]: Handled: _grouped_bnb4bit_ready fails the probe when any expert bias is None (and when bias.requires_grad), so it falls back to the per-expert loop instead of reaching torch.stack.

File: unsloth_zoo/temporary_patches/gpt_oss.py [hunk 7]
```diff
@@ -848,13 +849,133 @@ def __init__(self, config):
...
+        gate_up = gate_up + gate_up_bias[expert_ids]
+        gated = swiglu_torch_forward(gate_up, self.alpha, self.limit)
+        out = _base_grouped_mm(
+            gated, offsets, lambda: _stack(self.down_projs, down_qs), recompute)
+        out = out + down_bias[expert_ids]
+
+        weighted = out.to(torch.float32) * routing_weights[sorted_tokens, sorted_experts, None].to(torch.float32)
+        next_states = torch.zeros(num_tokens, self.hidden_size, dtype=torch.float32, device=device)
+        next_states.index_add_(0, sorted_tokens, weighted)
+        return next_states.view(batch_size, -1, self.hidden_size).to(hidden_states.dtype)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/gpt_oss.py]: The grouped path accumulates and combines in float32 and returns next_states.to(torch.float32), matching torch_native_forward training branch.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
### Final results

Setup: `unsloth/gpt-oss-20b-unsloth-bnb-4bit` LoRA SFT, bnb 4-bit experts, grouped `torch._grouped_mm` path vs the per-expert Python loop. The loop launches a dequantize and two matmuls per expert every layer even for experts that received no tokens; a profiled loop step was about 80 percent GPU-idle (2.27s CPU vs 0.61s CUDA, 33k kernel launches, 2062 per-expert 4-bit matmuls, 3140 dequantize calls).

| Seq length | per-expert loop tok/s | grouped tok/s | speedup |
|------------|-----------------------|---------------|---------|
| 4k | 2433 | 10194 | 4.2x |
| 8k | - | 11238 | - |

Per-step losses match the loop path to 0.0022 over 20 deterministic steps, so the grouped path is a throughput change only. It engages behind a readiness probe (grouped_mm support, no LoRA wrappers, frozen quantized weights); `UNSLOTH_GPTOSS_GROUPED=0` or `UNSLOTH_COMPILE_DISABLE=1` falls back to the per-expert loop.
